.github/workflows/build-and-test.yml

```diff
@@ -593,3 +593,113 @@ jobs:
+    runs-on: [self-hosted, xlarge]
+    steps:
+      - name: Download charm
+        uses: actions/download-artifact@v3
+        with:
+          name: charms
+          path: ~/artifacts/
+
+      - name: Checkout code
+        uses: actions/checkout@v3
+        with:
+          fetch-depth: 0
+
+      - name: Setup LXD
+        uses: canonical/setup-lxd@v0.1.1
+        with:
+          # pin lxd to LTS release.
+          channel: 5.21/stable
+
+      - name: Install dependencies
+        run: ./tests/scripts/ci_helpers.sh install_deps
+
+      - name: Install MicroCeph charm
+        run: |
+          date
+          mv ~/artifacts/microceph.charm ./microceph.charm
+          juju deploy ./tests/bundles/multi_node_juju_storage.yaml 
+          # wait for charm to bootstrap and OSD devices to enroll.
+          ./tests/scripts/ci_helpers.sh wait_for_microceph_bootstrap
+          date
+
+      - name: Show Juju status
+        run: |
+          set -eux
+          juju status
+
+      - name: Remove Unit microceph/2
+        run: |
+          set -x
+          disk_list=$(juju ssh microceph/leader -- "sudo microceph disk list --json")
+          osd_count=$(echo $disk_list | jq '.ConfiguredDisks[].location' | grep -c "\-2")
+          if [[ $osd_count -ne 3 ]] ; then
+            echo "Unexpected OSD count on node microceph/2 $osd_count"
+            exit 1
+          fi
+
+          juju remove-unit microceph/2 --no-prompt
+          # wait and check if the unit is still present.
+          for i in $(seq 1 40); do
+            res=$( ( juju status | grep -cF "microceph/2" ) || true )
+            if [[ $res -gt 0 ]] ; then
+              echo -n '.'
+              sleep 5
+            else
+              echo "Unit removed successfully"
+              break
+            fi
+          done
+          # fail if unit still present.
+          if [[ $res -gt 0 ]] ; then
+            echo "Unit still present"
+            juju status
+            exit 1
+          fi
+
+          # sleep for some time
+          sleep 1m
+
+          disk_list=$(juju ssh microceph/leader -- "sudo microceph disk list --json")
+          new_osd_count=$(echo $disk_list | jq '.ConfiguredDisks[].location' | grep -c "\-2" || true)
+          if [[ $new_osd_count -gt 0 ]] ; then
+            echo "Unexpected OSD count on node microceph/2 $osd_count"
+            exit 1
+          fi
+      
+      - name: Add another unit
+        run: |
+          set -x
+          juju add-unit microceph -n 1
+          sleep 60s
+          juju wait-for unit microceph/3 --query='workload-message=="(workload) charm is ready"' --timeout=20m
+          disk_list=$(juju ssh microceph/1 -- "sudo microceph disk list --json")
+          total_osd_count=$(echo $disk_list | jq '.ConfiguredDisks[].location' | grep -c "juju")
+          if [[ $total_osd_count -ne 9 ]] ; then
+            echo "Unexpected total OSD count on $osd_count"
+            exit 1
+          fi
```

Comment:
Feels like there would be opportunity for code dedup by factoring out a `check_osd_count($location)` helper function

Reply:
ack